### ISSUE TITLE
Tracing some ruby USDT probes causes an invalid BPF program error

### ISSUE THREAD

**Issue by goldshtn** [iovisor/bcc#722]
I'm trying to instrument ruby USDT probes with `trace`. Some of the probes work just fine, but there's a couple that I'm getting errors with. Sorry for the long intro, there's just a couple of steps needed to reproduce this.

First, grab the Ruby source code:

```
git clone https://github.com/ruby/ruby
cd ruby
autoconf
./configure --enable-dtrace
make
make install
```

Now you have ruby and irb in /usr/local/bin. `tplist` reports a bunch of probes:

```
tplist -l /usr/local/bin/ruby
```

The docs for the probes are [here](https://ruby-doc.org/core-2.1.0/doc/dtrace_probes_rdoc.html). Now run `/usr/local/bin/irb` and let's try instrumenting one that works fine:

```
trace 'u:ruby:object__create "%s %s %d", arg1, arg2, arg3' -p $(pidof irb)
```

This should work fine and print out the class name, file name, and line number every time you create an object in irb. Nice.

But the `method__entry` probe causes an invalid BPF program (same thing with `method__return`). The full output is below. From looking at the probe argument locations with `readelf`, I don't see anything particularly suspicious about `method__return` as opposed to `object__create`:

```
$ readelf -n /usr/local/bin/ruby | grep -A 2 ' method__entry' 
    Name: method__entry
    Location: 0x000000000018be34, Base: 0x00000000002af400, Semaphore: 0x00000000004ff0c8
    Arguments: 8@944(%rsp) 8@952(%rsp) 8@960(%rsp) -4@968(%rsp)
--
    Name: method__entry
    Location: 0x000000000018d921, Base: 0x00000000002af400, Semaphore: 0x00000000004ff0c8
    Arguments: 8@(%rsp) 8@8(%rsp) 8@16(%rsp) -4@24(%rsp)

$ readelf -n /usr/local/bin/ruby | grep -A 2 ' object__create' 
    Name: object__create
    Location: 0x0000000000092b2a, Base: 0x00000000002af400, Semaphore: 0x00000000004ff0de
    Arguments: 8@%rax 8@%r12 -4@12(%rsp)
```

Error output including BPF program source, this is from `method__entry`. I've also modified `trace` slightly so that it prints the USDT helpers as well:

```c
# ./trace.py 'u:ruby:method__entry "%s", arg1' -p $(pidof irb) -v

#include <linux/ptrace.h>
#include <linux/sched.h>        /* For TASK_COMM_LEN */


struct probe_method__entry_1_data_t
{
        u64 timestamp_ns;
        u32 pid;
        char comm[TASK_COMM_LEN];
        char v0[80];   



};

BPF_PERF_OUTPUT(probe_method__entry_1_events);



int probe_method__entry_1(struct pt_regs *ctx)
{

        u32 __pid = bpf_get_current_pid_tgid();
        if (__pid == 5943) { return 0; }



        if (!(1)) return 0;

        struct probe_method__entry_1_data_t __data = {0};
        __data.timestamp_ns = bpf_ktime_get_ns();
        __data.pid = bpf_get_current_pid_tgid();
        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
        u64 arg1;
        bpf_usdt_readarg(1, ctx, &arg1);

        if (arg1 != 0) {
                bpf_probe_read(&__data.v0, sizeof(__data.v0), (void *)arg1);
       }


        probe_method__entry_1_events.perf_submit(ctx, &__data, sizeof(__data));
        return 0;
}

#include <uapi/linux/ptrace.h>
static inline int _bpf_readarg_probe_method__entry_1_1(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(uint64_t)) return -1;
  switch(ctx->ip) {
  case 0x55f2d31a1e34ULL: { u64 __addr = ctx->sp + (944); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  case 0x55f2d31a3921ULL: { u64 __addr = ctx->sp + (0); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  }
  return -1;
}
static inline int _bpf_readarg_probe_method__entry_1_2(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(uint64_t)) return -1;
  switch(ctx->ip) {
  case 0x55f2d31a1e34ULL: { u64 __addr = ctx->sp + (952); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  case 0x55f2d31a3921ULL: { u64 __addr = ctx->sp + (8); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  }
  return -1;
}
static inline int _bpf_readarg_probe_method__entry_1_3(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(uint64_t)) return -1;
  switch(ctx->ip) {
  case 0x55f2d31a1e34ULL: { u64 __addr = ctx->sp + (960); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  case 0x55f2d31a3921ULL: { u64 __addr = ctx->sp + (16); uint64_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((uint64_t *)dest) = __res; } return 0;
  }
  return -1;
}
static inline int _bpf_readarg_probe_method__entry_1_4(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(int32_t)) return -1;
  switch(ctx->ip) {
  case 0x55f2d31a1e34ULL: { u64 __addr = ctx->sp + (968); int32_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((int32_t *)dest) = __res; } return 0;
  case 0x55f2d31a3921ULL: { u64 __addr = ctx->sp + (24); int32_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((int32_t *)dest) = __res; } return 0;
  }
  return -1;
}

bpf: Permission denied 
0: (bf) r6 = r1
1: (85) call 14
2: (67) r0 <<= 32
3: (77) r0 >>= 32
4: (15) if r0 == 0x1737 goto pc+59
 R0=inv32 R6=ctx R10=fp
5: (b7) r7 = 0
6: (7b) *(u64 *)(r10 -24) = r7
7: (7b) *(u64 *)(r10 -32) = r7
8: (7b) *(u64 *)(r10 -40) = r7
9: (7b) *(u64 *)(r10 -48) = r7
10: (7b) *(u64 *)(r10 -56) = r7
11: (7b) *(u64 *)(r10 -64) = r7
12: (7b) *(u64 *)(r10 -72) = r7
13: (7b) *(u64 *)(r10 -80) = r7
14: (7b) *(u64 *)(r10 -88) = r7
15: (7b) *(u64 *)(r10 -96) = r7
16: (7b) *(u64 *)(r10 -104) = r7
17: (7b) *(u64 *)(r10 -112) = r7
18: (7b) *(u64 *)(r10 -120) = r7
19: (85) call 5
20: (7b) *(u64 *)(r10 -128) = r0
21: (85) call 14
22: (63) *(u32 *)(r10 -120) = r0
23: (bf) r1 = r10
24: (07) r1 += -116
25: (b7) r2 = 16
26: (85) call 16
27: (79) r1 = *(u64 *)(r6 +128)
28: (18) r2 = 0xd31a3921
30: (1d) if r1 == r2 goto pc+10
 R0=inv R1=inv R2=inv R6=ctx R7=imm0 R10=fp
31: (18) r2 = 0xd31a1e34
33: (5d) if r1 != r2 goto pc+16
 R0=inv R1=inv R2=inv R6=ctx R7=imm0 R10=fp
34: (79) r3 = *(u64 *)(r6 +152)
35: (b7) r1 = 0
36: (7b) *(u64 *)(r10 -8) = r1
37: (07) r3 += 944
38: (bf) r7 = r10
39: (07) r7 += -8
40: (05) goto pc+4
45: (bf) r1 = r7
46: (b7) r2 = 8
47: (85) call 4
48: (79) r3 = *(u64 *)(r7 +0)
49: (15) if r3 == 0x0 goto pc+4
 R0=inv R3=inv R6=ctx R7=fp-8 R10=fp
50: (bf) r1 = r10
51: (07) r1 += -100
52: (b7) r2 = 80
53: (85) call 4
54: (18) r7 = 0xba873bc0
56: (85) call 8
57: (bf) r4 = r10
58: (07) r4 += -128
59: (bf) r1 = r6
60: (bf) r2 = r7
61: (bf) r3 = r0
62: (b7) r5 = 112
63: (85) call 25
64: (b7) r0 = 0
65: (95) exit

from 49 to 54: safe

from 33 to 50: R0=inv R1=inv R2=inv R6=ctx R7=imm0 R10=fp
50: (bf) r1 = r10
51: (07) r1 += -100
52: (b7) r2 = 80
53: (85) call 4
R3 !read_ok

Traceback (most recent call last):
  File "./trace.py", line 605, in run
    self._attach_probes()
  File "./trace.py", line 582, in _attach_probes
    self.bpf = BPF(text=self.program, usdt_contexts=usdt_contexts)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 209, in __init__
    usdt_context.attach_uprobes(self)
  File "/usr/lib/python2.7/site-packages/bcc/usdt.py", line 81, in attach_uprobes
    addr=addr, pid=pid)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 612, in attach_uprobe
    fn = self.load_func(fn_name, BPF.KPROBE)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 256, in load_func
    raise Exception("Failed to load BPF program %s: %s" % (func_name, errstr))
Exception: Failed to load BPF program probe_method__entry_1: Permission denied
```

Any thoughts would be appreciated. This is on FC23, 4.8.0-1.vanilla.knurd.1.fc23.x86_64, with bcc master.

**Comment by 4ast** [comment]
reproduced.
In the following:
```
        u64 arg1;
        bpf_usdt_readarg(1, ctx, &arg1);

        if (arg1 != 0) {
                bpf_probe_read(&__data.v0, sizeof(__data.v0), (void *)arg1);
```
arg1 can stay uninitialized at the time of bpf_probe_read,
so R3 is not inited and verifier complains

**Comment by goldshtn** [comment]
@4ast But that's highly bizarre, because the same thing exactly should happen with any other probe that touches strings (and therefore has the `argN != 0` test). However, it works perfectly fine with the `object__create` probe. What gives?

**Comment by goldshtn** [comment]
Oh, maybe it's the fact that the probe has two locations, so there is a slightly different codepath (the readarg helpers don't do exactly the same thing when there's just one location). I'll make sure. In any case, it looks like initializing the args before calling readarg is the easy fix.

**Comment by 4ast** [comment]
in case of 'object__create' arg1 is always initialized:
static inline int _bpf_readarg_probe_object__create_1_1(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(uint64_t)) return -1;
  *((uint64_t *)dest) = (uint64_t)ctx->r14;
  return 0;
}
so yes, two locations kinda caused it :)

**Comment by goldshtn** [comment]
But actually if the length check fails then `dest` is not initialized. Or is the JIT smart enough to realize that `len`is a constant and fold the test away?

**Comment by 4ast** [comment]
llvm knows that len is constant and optimizes it out.

**Comment by goldshtn** [comment]
Great. Indeed, just adding that little `= 0` fixes this issue (#725). We can haz ruby tracing 😀
